### ISSUE TITLE
Add box type constructor

### ISSUE THREAD

**Issue by andy-morris** [anoma/juvix#879]
The 'box' type `□ π A` will let you package up several uses of a value. Possibly most importantly, it'll let you return a value that can be used several times with a type like `sig add-with-two-copies : nat -2-> nat -2-> box 2 nat`.

## Core

In core, at the moment, □ will have to be a new primitive syntactic form. It can be a datatype but only once we have first class usages to parameterise by the π.

These rules are based on GRTT, but adjusted a bit to fit into a bidirectional system, and also because GRTT treats usages/quantities/grades/whatever differently.

- Type formation:
  ```
       0Γ ⊢ 0 A ⇐ ★ i
  ────────────────────────
     0Γ ⊢ 0 □ π A ⇐ ★ i
  ```
  As you can see this rule requires □ to checked in a zero context, like all types. If we decide to allow runtime-relevant types then we need to fix this rule up too, but currently we don't.

- Introduction:
  ```
       Γ ⊢ σπ s ⇐ A
  ────────────────────────
    Γ ⊢ σ 𝐛𝐨𝐱 s ⇐ □ π A
  ```
  I'm pretty sure the implementation will be the same way: checking a box will only require adjusting the usage and then checking the contents.

- Elimination:
  ```
            0Γ, 0 x : A ⊢ 0 B ⇐ ★ i
     Γ ⊢ σ e ⇒ □ π A     Γ, π x : A ⊢ σ s ⇐ B
  ──────────────────────────────────────────────
     Γ ⊢ 𝐥𝐞𝐭-𝐛𝐨𝐱 x = e 𝐫𝐞𝐭𝐮𝐫𝐧 B 𝐢𝐧 s ⇒ B[e/x]
  ```
  You might remember that Σ doesn't have an eliminator in core, in favour of using pattern-matching functions. But I think since first class usages are a longer way off it would be better to have an eliminator for box, since its datatype replacement is definitely not going to make it into the next release.

- Computation:
  ```
                   s ⟿ u    t ⟿ v
  ───────────────────────────────────────────────────
   𝐥𝐞𝐭-𝐛𝐨𝐱 x = (𝐛𝐨𝐱 t : □ π A) 𝐫𝐞𝐭𝐮𝐫𝐧 B 𝐢𝐧 s ⟿ u[v/x]
  ```
  similar to the existing β for functions taken from IGPON

## Frontend syntax

I don't have _too_ strong an opinion about the frontend syntax. I guess something like:

- Type: `box π A` for types (where `π` is either a nat literal or a name bound to `%Builtin.Omega`). I assume something fancy like `[π] A` clashes with list literals, and in any case for backcompat reasons it would probably be better to write it in a way that its eventual datatype replacement could also be.
- Intro: `Box s`, since our convention is to have type names lowercase and data constructors capitalised.
- Elim: same as in core? (Maybe `let Box x` instead of `let-box x`?) Once the unifier exists, allow the `return B` to be optional and fill it with a meta if it's missing.

**Comment by cwgoes** [comment]
I don't understand the bit on computation, can you clarify?

The erasure rules seem straightforward but does this affect execution at all?

**Comment by andy-morris** [comment]
Since values don't have type annotations (since they're only needed on redexes and top level defs), the computation can be simplified to something like

```
𝐥𝐞𝐭-𝐛𝐨𝐱 x = (𝐛𝐨𝐱 u : □ π A) 𝐫𝐞𝐭𝐮𝐫𝐧 B 𝐢𝐧 v ⟿ v[u/x]
```

fwiw the one above (and function β in IGPON) are like that because variables are elims, but the argument to a function and the body of 𝐥𝐞𝐭-𝐛𝐨𝐱 are terms, so when you do the substitution you need to carry the type annotations around at the same time, e.g.

```
𝐥𝐞𝐭-𝐛𝐨𝐱 f = (𝐛𝐨𝐱 (λx. x) : □ 1 (1 ℕ → ℕ)) 𝐫𝐞𝐭𝐮𝐫𝐧 Σ(_: ℕ) → Unit 𝐢𝐧 〈f 0, 〈〉〉
⟿
〈(λx. x : 1 ℕ → ℕ) 0, 〈〉〉 : Σ(_: ℕ) → Unit
```
where the λ term needs a type annotation to be the head of an application, and the pair constructor needs one because the whole expression is an elimination context too.

------

For erasure, I'm actually not sure what to do about values of type `□ 0 A`. Perhaps the erasure can translate `□ 0 _` and `box _ : □ 0 _` to unit type/value, and `□ π A` and `box s : □ π A` to `A`/`s : A` when `π ≠ 0`?

Anyway once we have a satisfactory answer to that, no, execution shoudn't be affected, since we don't actually *do* anything at runtime (reference counting, or whatever) when we have multple uses of a value.

**Comment by andy-morris** [comment]
superseded by records.
